models/classes/Lists/DataAccess/Repository/ParentPropertyListCachedRepository.php

```diff
@@ -26,51 +26,79 @@
-            foreach ($this->getParentPropertyListRepository()->findAllUris($options) as $propertyUri) {
-                $this->getCache()->delete(sprintf(self::CACHE_MASK, $propertyUri, $options['listUri']));
-            }
-
-            return;
+            $this->removeUsingListUri($options['listUri']);
+        }
+        $childListUris = $this->getDependencyRepository()->findChildListUris([
+                            'parentListUri' => $options['listUri']
+                        ]);
+        foreach ($childListUris as $uri) {
+            $this->removeUsingListUri($uri);
         }
-
-        $this->getCache()->delete(sprintf(self::CACHE_MASK, $options['propertyUri'], $options['listUri']));
     }
 
     public function findAllUris(array $options): array
     {
-        //@TODO Rework cache considering only listUri...
         /** @var core_kernel_classes_Property $property */
-//        $property = $options['property'];
-//        $listUri = $options['listUri'] ?? $property->getRange()->getUri();
-//
-//        $cacheKey = sprintf(self::CACHE_MASK, $property->getUri(), $listUri);
-//
-//        if ($this->getCache()->has($cacheKey)) {
-//            return $this->getCache()->get($cacheKey);
-//        }
+        $property = $options['property'] ?? null;
+        if (!$property) {
+            return $this->getParentPropertyListRepository()->findAllUris($options);
+        }
+        $listUri = $options['listUri'] ?? $property->getRange()->getUri();
+        $cacheKey = sprintf(self::CACHE_MASK, $property->getUri(), $listUri);
+        $listCacheKey = sprintf(self::LIST_CACHE_MASK, $listUri);
+        $currentValues = [];
+        $listCacheValues = [$cacheKey];
+
+        if ($this->getCache()->has($listCacheKey)) {
+            $currentValues = $this->getCache()->get($listCacheKey);
+            $listCacheValues = array_unique(array_merge($currentValues, $listCacheValues));
+        }
+
+        if ($currentValues !== $listCacheValues) {
+            $this->getCache()->set($listCacheKey, $listCacheValues);
+        }
+
+        if ($this->getCache()->has($cacheKey)) {
+            return $this->getCache()->get($cacheKey);
+        }
 
         $uris = $this->getParentPropertyListRepository()->findAllUris($options);
 
-        //$this->getCache()->set($cacheKey, $uris);
+        $this->getCache()->set($cacheKey, $uris);
 
         return $uris;
     }
 
+    private function removeUsingListUri(string $uri): void
+    {
+        $listCacheKey = sprintf(self::LIST_CACHE_MASK, $uri);
+        if ($this->getCache()->has($listCacheKey)) {
+            $listCacheValues = $this->getCache()->get($listCacheKey);
+            foreach ($listCacheValues as $value) {
+                [$key, $propertyUri, $listUri]  = explode('-', $value);
+                $this->getCache()->delete(sprintf(self::CACHE_MASK, $propertyUri, $listUri));
+            }
+            $this->getCache()->delete($listCacheKey);
+        }
```

Comment:
Please add blank line to help people read the code easily.

```suggestion
        $listCacheKey = sprintf(self::LIST_CACHE_MASK, $uri);
        
        if ($this->getCache()->has($listCacheKey)) {
            $listCacheValues = $this->getCache()->get($listCacheKey);
            
            foreach ($listCacheValues as $value) {
                [$key, $propertyUri, $listUri]  = explode('-', $value);
                $this->getCache()->delete(sprintf(self::CACHE_MASK, $propertyUri, $listUri));
            }
            
            $this->getCache()->delete($listCacheKey);
        }
```

Reply:
Done